<a href="https://colab.research.google.com/github/QuirozAlaniz/Datos_duplicados/blob/main/Actividad_Calificada_07_Datos_Duplicados_MATRICULA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | 25 de Septiembre de 2026 |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** |Francisco Daniel Quiroz Alaniz |
| **Matrícula** |269723 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [554]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [555]:
# Escribe tu código aquí
hoteles = pd.DataFrame(hoteles)
duplicados_first = hoteles.duplicated(subset = None, keep = "first").sum()
duplicados_last = hoteles.duplicated(subset = None, keep = "last").sum()
duplicados_false = hoteles.duplicated(subset = None, keep = False).sum()
print(f'Duplicados con el parámetro "first": {duplicados_first}, son los duplicados pero sin tomar en cuenta la primera ocurrencia')
print()
print(f'Duplicados con el parámetro "Last": {duplicados_last} son los duplicados pero sin tomar en cuenta la última ocurrencia')
print()
print(f'Duplicados con el parámetro False: {duplicados_false}, son todos los dulicados incluyendo todas las ocurrencias')

Duplicados con el parámetro "first": 31994, son los duplicados pero sin tomar en cuenta la primera ocurrencia

Duplicados con el parámetro "Last": 31994 son los duplicados pero sin tomar en cuenta la última ocurrencia

Duplicados con el parámetro False: 40165, son todos los dulicados incluyendo todas las ocurrencias


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [556]:
# Escribe tu código aquí
proporcion = hoteles.duplicated().mean()
total_filas = len(hoteles)
duplicadas = hoteles.duplicated(subset = None, keep = "first").sum()
prop = (duplicadas/total_filas)
print(f' Filas duplicadas respecto al total: {prop:.2%}')
print()
repetidos = hoteles.value_counts()
print("Frecuencia de cada fila completa")
print(repetidos)

 Filas duplicadas respecto al total: 26.80%

Frecuencia de cada fila completa
hotel         is_canceled  lead_time  arrival_date_year  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  meal  country  market_segment  distribution_channel  is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  reserved_room_type  assigned_room_type  booking_changes  deposit_type  agent  company  days_in_waiting_list  customer_type    adr     required_car_parking_spaces  total_of_special_requests  reservation_status  reservation_status_date
City Hotel    0            256        2016               October             43                        16                         2                        3                     2       0.0       0       BB    DEU      Online TA       TA/TO                 0                  0                       0                               A                   A     

### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [557]:
# Escribe tu código aquí
hoteles2 = hoteles.copy()                              #Hacemos una copia del dataset original y lo guardamos en "hoteles2"

In [558]:
hoteles2["es_duplicado"] = hoteles.duplicated(subset = None, keep = "first")                 #Agregamos la columna booleana a los duplicados sin contar primeras ocurrencias
agrupacion = hoteles2[["hotel", "es_duplicado"]].groupby("hotel", as_index = False).agg({"es_duplicado":"sum"})     #El "as_index = False" evita que la columna hotel
                                                                                                                    #se vuelva indice
duplicados_city = agrupacion.iloc[0, 1]                                                                             #Accedemos al numero de duplicados del City Hotel
duplicados_resort = agrupacion.iloc[1, 1]                                                                           #Accedemos al número de duplicados del Resort Hotel
print(f'Los duplicados del City Hotel son: {duplicados_city}')
print()
print(f'Los duplicados del Resort Hotel son: {duplicados_resort}')

reservas_city = hoteles2[hoteles2["hotel"] == "City Hotel"]                                                  #Obtenemos el total de reservas de City Hotel

reservas_resort = hoteles2[hoteles2["hotel"] == "Resort Hotel"]                                                #Obtenemos el total de reservas de Resort Hotel

print()
print(f'El porcentaje de registros duplicados en City Hotel es de {duplicados_city/len(reservas_city):.2%}')
print()
print(f'El porcentaje de registros duplicados en Resort Hotel es de {duplicados_resort/len(reservas_resort):.2%}')

Los duplicados del City Hotel son: 25902

Los duplicados del Resort Hotel son: 6092

El porcentaje de registros duplicados en City Hotel es de 32.65%

El porcentaje de registros duplicados en Resort Hotel es de 15.21%


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**
\
El problema de duplicados de "City Hotel" es grave porque los duplicados llegan a ser el $32.65\%$, por lo que podríamos considerarlos datos insuficientes.
\
El problema se reduce en "Resort Hotel" porque llega a $15.21\%$, lo cual no es tan grave como el anterior pero sí se podría llegar a trabajar sin problemas con esa cantidad de información al ser $85\%$ la disponible.  

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [559]:
# Escribe tu código aquí
hoteles2 = hoteles2.drop(columns = ["es_duplicado"])        #Le quitamos la columna "es_duplicado" que le había puesto para que nos quede como el original
hoteles2["duplicados_false"]= hoteles2.duplicated(subset = None, keep = False) #Agregamos la columna "duplicados_false" que muestra todos los datos duplicados incluyendo todas
                                                                             #las ocurrencias con "True".

duplicados = hoteles2[hoteles2["duplicados_false"] == True]                 #hacemos una variable que tiene solamente las filas duplicadas, incluidas las primeras ocurrencias

columnas = ["hotel", "lead_time", "arrival_date_year", "arrival_date_month", "arrival_date_day_of_month", "adults", "country",
            "market_segment","deposit_type","adr", "reservation_status", "duplicados_false"]          #definimos una lista de columnas, incluyendo la que acabamos de hacer

duplicados_ordenados = duplicados.sort_values(by = "lead_time")              #Elegí ordenar por lead_time para tenerlos ordenados del más reciente al más antiguo, es una
                                                                             #manera de prestar atención a los que van a ocurrir primero
print(duplicados_ordenados[columnas].head(12))                               #imprimimos las 12 primeras filas de las columnas elegidas con los duplicados


               hotel  lead_time  arrival_date_year arrival_date_month  \
63326     City Hotel          0               2017           February   
63324     City Hotel          0               2017           February   
105321    City Hotel          0               2017           February   
15453   Resort Hotel          0               2015               July   
15452   Resort Hotel          0               2015               July   
105346    City Hotel          0               2017           February   
105345    City Hotel          0               2017           February   
47243     City Hotel          0               2016           February   
47248     City Hotel          0               2016           February   
47036     City Hotel          0               2016           February   
47034     City Hotel          0               2016           February   
39360   Resort Hotel          0               2017             August   

        arrival_date_day_of_month  adults country 

### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [560]:
# Escribe tu código aquí
hoteles2 = hoteles2.drop(columns = ["duplicados_false"])          #Regresamos el dataset a su composición original
numero_repetidos = hoteles2.value_counts(dropna = False)          #Contamos el número de repeticiones de las filas únicas del dataframe original y la hacemos Serie
                                                                  #El dropna = False es para que no ignore a los na
combinacion = numero_repetidos.idxmax()                           #Metemos en la variable combinación los índices de la serie "numeros_repetidos"

print(f'La frecuencia de cada repetición es la siguiente: \n {numero_repetidos.value_counts().head(10)}')
print()
print(f'La fila que viene más veces repetida contiene {numero_repetidos.max() } repeticiones y su combinación es: \n {combinacion}')

La frecuencia de cada repetición es la siguiente: 
 count
1     79225
2      5034
3       969
4       497
5       275
6       171
10      123
7       115
8       103
20       97
Name: count, dtype: int64

La fila que viene más veces repetida contiene 180 repeticiones y su combinación es: 
 ('City Hotel', np.int64(1), np.int64(277), np.int64(2016), 'November', np.int64(46), np.int64(7), np.int64(1), np.int64(2), np.int64(2), np.float64(0.0), np.int64(0), 'BB', 'PRT', 'Groups', 'TA/TO', np.int64(0), np.int64(0), np.int64(0), 'A', 'A', np.int64(0), 'Non Refund', np.float64(nan), np.float64(nan), np.int64(0), 'Transient', np.float64(100.0), np.int64(0), np.int64(0), 'Canceled', '2016-04-04')


### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [561]:
# Escribe tu código aquí
hoteles2["duplicados_false"]= hoteles2.duplicated(subset = None, keep = False)
duplicados = hoteles2[hoteles2["duplicados_false"] == True]                          #Creamos un dataframe con los datos que tengan solamente los duplicados

print(f'Proporciones de cada segumento de mercado del dataset original: \n {hoteles["market_segment"].value_counts(normalize = True)}')    #imprimimos las proporciones que representan cada categoría de la columna market_total
                                                                                                                                            #con respecto al número de filas del total.
print()
print(f'Proporciones de cada segmento de mercado del dataset de duplicados: \n {duplicados["market_segment"].value_counts(normalize = True) }')    #imprimimos el resultado
print()
print("-------------------------------------------------------------------------------------------")
print(f'Proporciones de cada tipo de depósito del dataset original : \n {hoteles["deposit_type"].value_counts(normalize = True)}')
print()
print(f'Proporción de cada tipo de depósito del dataset de duplicados: \n {duplicados["deposit_type"].value_counts(normalize = True)}')
print()
print("-------------------------------------------------------------------------------------------")
print(f'Proporción de cada tipo de estatus de cancelación del dataset original: \n {hoteles["is_canceled"].value_counts(normalize = True)}')
print()
print(f'Proporción de cada tipo de estatus de cancelación del dataset de duplicados: \n {duplicados["is_canceled"].value_counts(normalize = True)}')


Proporciones de cada segumento de mercado del dataset original: 
 market_segment
Online TA        0.473046
Offline TA/TO    0.202856
Groups           0.165935
Direct           0.105587
Corporate        0.044350
Complementary    0.006223
Aviation         0.001985
Undefined        0.000017
Name: proportion, dtype: float64

Proporciones de cada segmento de mercado del dataset de duplicados: 
 market_segment
Groups           0.416109
Offline TA/TO    0.303996
Online TA        0.205876
Corporate        0.036699
Direct           0.034981
Complementary    0.001867
Aviation         0.000473
Name: proportion, dtype: float64

-------------------------------------------------------------------------------------------
Proporciones de cada tipo de depósito del dataset original : 
 deposit_type
No Deposit    0.876464
Non Refund    0.122179
Refundable    0.001357
Name: proportion, dtype: float64

Proporción de cada tipo de depósito del dataset de duplicados: 
 deposit_type
No Deposit    0.641678
Non 

### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

_Escribe aquí tu respuesta._
\
a) Las filas duplicadas probablemente sean errores de registro de sistema, debido a que en todas las 32 columnas son idénticas, lo cual es muy improbable.
\
b) Por default duplicated() cuando compara dos celdas que tienen NaN son tomados en cuenta como iguales. Esto implica que si en realidad esos NaN son omisiones de registro accidentales, pueden provocar que registros se consideren duplicados cuando no lo serían si esas celdas con NaN tuvieran el registro correcto que pudiera ser diferente.  

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [562]:
# Escribe tu código aquí
hoteles_original = hoteles2.drop("duplicados_false", axis = 1)
print(f'# de filas del dataframe original: {hoteles_original.shape[0]} ')
hoteles_original["duplicados_first"] = hoteles_original.duplicated(subset = None, keep = "first")  #Agregamos la columna booleana, donde los True son los duplicados
hoteles_limpio = hoteles_original[ hoteles_original["duplicados_first"] == False ]
hoteles_limpio.reset_index()                                                                       #Reestablecemos el índice
print(f'# de filas del dataframe limpio: {hoteles_limpio.shape[0]} ')
print(f'Se eliminaron {hoteles_original.shape[0]-hoteles_limpio.shape[0]} registros')

# de filas del dataframe original: 119390 
# de filas del dataframe limpio: 87396 
Se eliminaron 31994 registros


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [563]:
# Escribe tu código aquí
#Hacemos los cálculos para el dataframe original.

hoteles = hoteles_original.drop("duplicados_first", axis = 1)

agrupacion = hoteles.groupby(by = "hotel").size()

cancelacion = hoteles.groupby(by = "hotel")["is_canceled"].mean()*100

tarifa_promedio = hoteles.groupby("hotel")["adr"].mean()

#Hacemos los cálculos para el dataframe limpio.
agrupacion_limpio = hoteles_limpio.groupby("hotel").size()

cancelacion_limpio = hoteles_limpio.groupby("hotel")["is_canceled"].mean()*100

tarifa_promedio_limpio = hoteles_limpio.groupby("hotel")["adr"].mean()

#Creamos las tablas como dataframes para mostrar los resultados
resumen_original = pd.DataFrame({"hotel":agrupacion.index, "tasa_cancelacion":cancelacion.values, "tarifa_diaria_promedio": tarifa_promedio.values})
print(f'El resumen del dataframe original es: \n {resumen_original}')
print()
resumen_limpio = pd.DataFrame({"hotel": agrupacion_limpio.index, "tasa_cancelacion":cancelacion_limpio.values , "tarifa_diaria_promedio":tarifa_promedio_limpio.values })
print(f'El resumen del dataframe limpio es: \n {resumen_limpio}')

El resumen del dataframe original es: 
           hotel  tasa_cancelacion  tarifa_diaria_promedio
0    City Hotel         41.726963              105.304465
1  Resort Hotel         27.763355               94.952930

El resumen del dataframe limpio es: 
           hotel  tasa_cancelacion  tarifa_diaria_promedio
0    City Hotel         30.038557              110.985944
1  Resort Hotel         23.480923               99.025346


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [564]:
# Escribe tu código aquí
resumen_original["version"] = ["original", "original"]
resumen_limpio["version"] = ["sin duplicados", "sin duplicados"]


In [565]:
comparacion = pd.concat([resumen_original, resumen_limpio], ignore_index = True)       #Ponemos un dataframe encima del otro e ignoramos el índice.
comparacion = comparacion.sort_values(by = "hotel").reset_index(drop = True)           #Reordenamos por hotel y reseteamos el índice.
print(comparacion)

          hotel  tasa_cancelacion  tarifa_diaria_promedio         version
0    City Hotel         41.726963              105.304465        original
1    City Hotel         30.038557              110.985944  sin duplicados
2  Resort Hotel         27.763355               94.952930        original
3  Resort Hotel         23.480923               99.025346  sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

_Escribe aquí tu respuesta._
\
a) Con base en la tabla "comparación" generada, podemos ver que la tasa de cancelación sin duplicados en el City Hotel pasa de $41 \%$ a $30 \%$, son 11 puntos porcentuales de diferencia cuando quitamos los duplicados dejando las primeras ocurrencias, mientras que Resort Hotel solamente bajó de 27.7% a 23.4%.
\
\
b) Si un modelo se entrenara para predecir is_canceled con los datos originales, los duplicados causarían un problema porque, si el modelo considera frecuencia como "peso", entonces ponderaría de más un registro y su capacidad de generalización se vería sesgada.

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [566]:
# Escribe tu código aquí
hoteles_limpio = hoteles_limpio.copy()
                                                                                             #Para verificar que no hay datos duplicados, agregamos una columna booleana:
hoteles_limpio["duplicados"] = hoteles_limpio.duplicated(subset = None, keep = False)        #Agregamos la columna de duplicados incluyendo la primera ocurrencia
print(hoteles_limpio["duplicados"].value_counts())                                           #Contamos cuántas ocurrencias tiene cada grupo

indices = pd.DataFrame(hoteles_limpio.index)                                                 #Obtenemos los índices del dataframe hoteles_limpio y lo guardamos en "indices"
print(indices.duplicated(subset = None, keep = False).value_counts())                        #Revisamos si hay duplicados, como no hay duplicados entonces son valores únicos.

duplicados
False    87396
Name: count, dtype: int64
False    87396
Name: count, dtype: int64


Como la salida arroja solamente "False", significa que no hay registros repetidos.

### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [567]:
hoteles_limpio = hoteles_limpio.drop(columns = ["duplicados_first", "duplicados"]) #Eliminamos las columnas de apoyo que le había puesto en el ejercicio anterior

In [568]:
# Escribe tu código aquí

columnas = list(hoteles_limpio.columns)                                                      #hacemos una lista de todas las columnas que tiene hoteles_limpio
columnas.remove("reservation_status_date")                                                   #Le quitamos la columna de "reservation_status_date" al dataframe que hicimos.
hoteles_limpio["sobrantes"] = hoteles_limpio.duplicated(subset = columnas, keep = "first")   #Hacemos el booleano con todos los duplicados sin tomar en cuenta "reservation_Datatus_date"
                                                                                             #tomando en cuenta todas las ocurrencias.
filas_sobrantes = hoteles_limpio[hoteles_limpio["sobrantes"] == True]                        #Filtramos las filas sobrantes con todas las columnas
cantidad_sobrantes = filas_sobrantes["sobrantes"].sum()                                                   #contamos cuántas filas sobran
print(f'Solución 1: \n Si tomamos en cuenta todas las columnas menos "reservation_status_date", entonces sobrarían {cantidad_sobrantes} filas')
print()
columnas_mostrar = ["hotel", "lead_time", "arrival_date_year", "arrival_date_month", "arrival_date_day_of_month", "adr", "reservation_status", "reservation_status_date"]
print(f'Solución 2: \n Algunas de las filas sobrantes son: \n {filas_sobrantes[columnas_mostrar].head(5)}' )
print()


print(f' Solución 3: \n El estatus en esas filas es: \n {filas_sobrantes["reservation_status"].value_counts()}')




Solución 1: 
 Si tomamos en cuenta todas las columnas menos "reservation_status_date", entonces sobrarían 245 filas

Solución 2: 
 Algunas de las filas sobrantes son: 
              hotel  lead_time  arrival_date_year arrival_date_month  \
2019  Resort Hotel         81               2015          September   
3066  Resort Hotel         41               2015           November   
3165  Resort Hotel         36               2015           November   
3321  Resort Hotel         66               2015           December   
3328  Resort Hotel         66               2015           December   

      arrival_date_day_of_month    adr reservation_status  \
2019                         27  66.29           Canceled   
3066                         12  52.20           Canceled   
3165                         20  64.00           Canceled   
3321                          5  52.00           Canceled   
3328                          5  64.00           Canceled   

     reservation_status_date  
2019  

**Respuesta 4.2:**

_Escribe aquí tu respuesta._
\
La respuesta corta es que no dejaría esas filas debido a que ya están canceladas.
\
De hecho el que estén canceladas nos arroja información adicional, pues quizá fueron errores de registros y precisamente por eso fueron canceladas y no tiene caso mantenerlas.
\
Pero antes de eliminarlas indagaría más con el proveedor de los datos y con los expertos en el negocio para ver si esas cancelaciones hicieron otra reserva que reemplazó esta cancelada.

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

_Escribe aquí tu respuesta._
\
**Reporte de análisis de duplicados**
\
**Resultado**: Filas totales : 119,390, Filas únicas: 87,396, Duplicados: 31,994
\
**Criterio para eliminar filas**: el criterio utilizado es que todas las filas que coincidan exactamente en todas las columnas se considerarán duplicados.
\
**Regla de conservación**: Eliminamos los duplicados dejando solamente las primeras ocurrencias. Es reelevante para que no se pierda ningún registro necesario.
\
**Supuestos y limitaciones**: Suponemos que los duplicados son registros que no aportan valor, mientras que si coinciden en todas las columnas a excepción de al menos una, se deja porque sí aporta información adicional.
\
**Columnas totales**: 32


---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

_Escribe aquí tu respuesta._
\
1. Una situación en la que no conviene eliminar un duplicado de un dataset es cuando los registros duplicados correspondan a dos eventos iguales, por ejemplo en una tabla donde se registren los sueldos de empleados, puede darse el caso de que hay dos o más personas que tienen el mismo nombre y salario.
\
2. Le pediría la hora del registro de cada fila para poder compara en qué momento exacto se hizo cada una. Si se hicieron con una diferencia de segundos o minutos, es posible que haya sido un dobleteo por débil señal de internet o simplemente dieron "doble click" por accidente.
\
3. El método duplicated().sum() == 0 no garantiza que no haya duplicados porque no dice nada de los duplicados parciales ni toma en cuenta ni detecta los errores de captura.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.